symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Optimization.hs

```diff
@@ -0,0 +1,93 @@
+module ZkFold.Symbolic.Compiler.ArithmeticCircuit.Optimization (
+        optimize,
+        isLinear,
+        toConstVar
+    ) where
+
+import           Data.Bool                                               (bool)
+import           Data.Functor.Rep                                        (Representable (..))
+import           Data.Map                                                hiding (drop, foldl, foldr, map, null, splitAt,
+                                                                          take)
+import qualified Data.Map.Internal                                       as M
+import qualified Data.Map.Monoidal                                       as MM
+import qualified Data.Set                                                as S
+import           Numeric.Natural                                         (Natural)
+import           Prelude                                                 hiding (Num (..), drop, length, product,
+                                                                          splitAt, sum, take, (!!), (^))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial   (Mono (..))
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial (Poly (..), polynomial)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+
+--------------------------------- High-level functions --------------------------------
+
+-- | Replaces linear polynoms of the form
+-- @(fromConstant k) * (NewVar nV) + (fromConstant c)@
+-- with a constant variable @fromConstant $ negate c // k@ in an arithmetic circuit
+-- and replaces variable with a constant in witness
+--
+optimize ::
+  ( Arithmetic a, Ord (Rep i)) =>
+  ArithmeticCircuit a p i o  -> ArithmeticCircuit a p i o
+optimize ac = case [toConstVar p | (_, p) <- toList (acSystem ac), isLinear p] of
+    [] -> ac
+    vs -> optimize $ foldr optimize1 ac vs
+
+optimize1 :: forall a p i o. ( Arithmetic a, Ord (Rep i))
+  => (SysVar i , a )-> ArithmeticCircuit a p i o -> ArithmeticCircuit a p i o
+optimize1 (v, k) ac = case v of
+  NewVar nv ->  ac {
```

Comment:
`acOutput` might change as well if one of the outputs turns out to be a constant